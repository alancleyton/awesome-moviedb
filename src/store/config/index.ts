import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

import rootReducers from './root-reducers';
import rootSagas from './root-sagas';

/**
 * Creates a store config with reducers and sagas, returns a store
 */
const sagaMiddlewareOptions = {};

const sagaMiddleware = createSagaMiddleware(sagaMiddlewareOptions);

export const store = configureStore({
  reducer: rootReducers,
  middleware: [sagaMiddleware],
});

// Run saga middleware to make redux-sagas work
sagaMiddleware.run(rootSagas);

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;

// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
