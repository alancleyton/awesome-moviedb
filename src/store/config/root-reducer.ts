import { combineReducers } from '@reduxjs/toolkit';

import moviesReducer from '../movies/reducers';
import searchReducer from '../search/reducers';

/**
 * Root reducer with all reducers combined, place the reducers here
 */
const rootReducer = combineReducers({
  movies: moviesReducer,
  search: searchReducer,
});

export default rootReducer;
