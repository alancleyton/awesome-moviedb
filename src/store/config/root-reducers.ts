import moviesReducer from '../movies/reducers';

/**
 * Root reducer with all reducers combined, place the reducers here
 */
const rootReducers = {
  movies: moviesReducer,
};

export default rootReducers;
