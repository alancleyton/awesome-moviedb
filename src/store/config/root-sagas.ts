import { all } from 'redux-saga/effects';
import moviesSagas from '../movies/sagas';
import searchSagas from '../search/sagas';

/**
 * Root sagas with all sagas combined, place the sagas here
 */
export default function* rootSagas() {
  yield all([moviesSagas(), searchSagas()]);
}
