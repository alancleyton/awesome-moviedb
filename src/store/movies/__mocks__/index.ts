import { Movie } from '../types';

export const movies: Movie[] = [
  {
    adult: false,
    backdrop_path: '/xFYpUmB01nswPgbzi8EOCT1ZYFu.jpg',
    genre_ids: [28, 18, 12],
    id: 980489,
    original_language: 'en',
    original_title: 'Gran Turismo',
    overview:
      'Baseado na história de Jann Mardenborough, e relata a trajetória de um jogador de videogame que utilizou suas habilidades nos simuladores para se tornar um piloto profissional de verdade.',
    popularity: 3389.624,
    poster_path: '/qU60nhBRbKU23gIGrZi2CvUj6MQ.jpg',
    release_date: '2023-08-09',
    title: 'Gran Turismo: De Jogador a Corredor',
    video: false,
    vote_average: 8,
    vote_count: 763,
  },
];

export const error = {
  success: false,
  status_code: 34,
  status_message: 'The resource you requested could not be found.',
};
