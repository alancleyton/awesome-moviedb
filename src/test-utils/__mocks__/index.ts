import { Media } from '@/types/media';

export const searchResultMock = [
  {
    backdrop_path: '/vIgyYkXkg6NC2whRbYjBD7eb3Er.jpg',
    id: 580489,
    title: 'Venom: Tempo de Carnificina',
    original_title: 'Venom: Let There Be Carnage',
    overview:
      'Em busca de seu próximo furo de reportagem, o jornalista Eddie Brock consegue uma entrevista exclusiva com Cletus Kasady, um assassino no corredor da morte, que descobre o segredo de Eddie e se torna o hospedeiro de Carnificina, um simbionte aterrorizante e ameaçador. Eddie e Venom precisam superar sua relação atribulada e trabalhar em conjunto para derrotar Carnificina.',
    poster_path: '/h5UzYZquMwO9FVn15R2eK2itmHu.jpg',
    media_type: 'movie',
    adult: false,
    original_language: 'en',
    genre_ids: [878, 28, 12],
    popularity: 495.28,
    release_date: '2021-09-30',
    video: false,
    vote_average: 6.798,
    vote_count: 10204,
  },
  {
    backdrop_path: null,
    id: 504284,
    title: 'Carro',
    original_title: 'Carro',
    overview:
      'An undocumented Brazilian immigrant living in the Boston area decides to buy a car in an effort to better his life before returning home.',
    poster_path: '/3Ww0wNwgihuOEQasZLMLulrBXi7.jpg',
    media_type: 'movie',
    adult: false,
    original_language: 'pt',
    genre_ids: [18],
    popularity: 0.367,
    release_date: '2018-03-10',
    video: false,
    vote_average: 0.0,
    vote_count: 0,
  },
  {
    backdrop_path: null,
    id: 1241824,
    title: 'Carly',
    original_title: 'Carly',
    overview:
      'In a dystopian world where the only thing that matters is your "likeability", how far would you go to protect the future of the ones you love? Perhaps a guest spot at a high ratings TV show? What could go wrong… Welcome to the Cupcake Factory!',
    poster_path: null,
    media_type: 'movie',
    adult: false,
    original_language: 'en',
    genre_ids: [878],
    popularity: 0.351,
    release_date: '2023-10-13',
    video: false,
    vote_average: 0.0,
    vote_count: 0,
  },
  {
    id: 10959,
    name: 'Shia LaBeouf',
    original_name: 'Shia LaBeouf',
    media_type: 'person',
    adult: false,
    popularity: 41.834,
    gender: 2,
    known_for_department: 'Acting',
    profile_path: '/6JvHrMR4IYc8EMG0dJEigQNzbHU.jpg',
    known_for: [
      {
        backdrop_path: '/1mBRoDVPPJ4T7bDleBBaYvGUYUc.jpg',
        id: 1858,
        title: 'Transformers',
        original_title: 'Transformers',
        overview:
          'O destino da humanidade está em jogo quando duas raças de robôs, os Autobots e os vilões Decepticons, chegam à Terra. Os robôs possuem a capacidade de se transformarem em diferentes objetos mecânicos enquanto buscam a chave do poder supremo com a ajuda do jovem Sam.',
        poster_path: '/ev7IMfMXRbddGLVw7xPL2PHpkWR.jpg',
        media_type: 'movie',
        adult: false,
        original_language: 'en',
        genre_ids: [12, 878, 28],
        popularity: 24.542,
        release_date: '2007-06-27',
        video: false,
        vote_average: 6.809,
        vote_count: 11215,
      },
      {
        backdrop_path: '/zbdoCe1meTyy0v1fsIxZ9Q0ef3H.jpg',
        id: 8373,
        title: 'Transformers: A Vingança dos Derrotados',
        original_title: 'Transformers: Revenge of the Fallen',
        overview:
          "Passaram-se dois anos desde que Sam Witwicky e os Autobots salvaram a raça humana e tentam ter uma vida normal: Sam está começando a faculdade, enquanto os Autobots trabalham com uma organização militar secreta e tentam construir uma casa para eles na Terra. Porém, um antigo Decepticon conhecido como 'The Fallen' aparece para se vingar, então Sam e sua namorada Mikaela precisam desvendar a história dos Transformes e achar um caminho para derrotá-lo.",
        poster_path: '/xggdkMUwiY4zwlSXbp7AqPxAvQt.jpg',
        media_type: 'movie',
        adult: false,
        original_language: 'en',
        genre_ids: [878, 28, 12],
        popularity: 27.321,
        release_date: '2009-06-19',
        video: false,
        vote_average: 6.195,
        vote_count: 8456,
      },
      {
        backdrop_path: '/k6gUa0v8dgYI79OrdiwKVW7AdzL.jpg',
        id: 38356,
        title: 'Transformers: O Lado Oculto da Lua',
        original_title: 'Transformers: Dark of the Moon',
        overview:
          'Sam Witwicky e sua nova namorada, Carly, entram na briga quando os Decepticons renovam sua guerra contra os Autobots. Optimus Prime acredita que ressuscitar o antigo Transformer Sentinel Prime, ex-líder dos Autobots, pode levar à vitória. Essa decisão tem consequências devastadoras e a guerra parece pender a favor dos Decepticons, causando um conflito épico em Chicago.',
        poster_path: '/1B1o592zAr0pny4qPGceT56Ug2m.jpg',
        media_type: 'movie',
        adult: false,
        original_language: 'en',
        genre_ids: [28, 878, 12],
        popularity: 25.2,
        release_date: '2011-06-28',
        video: false,
        vote_average: 6.207,
        vote_count: 8265,
      },
    ],
  },
] as Media[];

export const moviesMock = [
  {
    adult: false,
    backdrop_path: '/xFYpUmB01nswPgbzi8EOCT1ZYFu.jpg',
    genre_ids: [28, 18, 12],
    id: 980489,
    original_language: 'en',
    original_title: 'Gran Turismo',
    overview:
      'Baseado na história de Jann Mardenborough, e relata a trajetória de um jogador de videogame que utilizou suas habilidades nos simuladores para se tornar um piloto profissional de verdade.',
    popularity: 3389.624,
    poster_path: '/qU60nhBRbKU23gIGrZi2CvUj6MQ.jpg',
    release_date: '2023-08-09',
    title: 'Gran Turismo: De Jogador a Corredor',
    video: false,
    vote_average: 8,
    vote_count: 763,
  },
] as Media[];

export const errorMock = {
  success: false,
  status_code: 34,
  status_message: 'The resource you requested could not be found.',
};
