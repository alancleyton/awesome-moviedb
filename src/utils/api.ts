import axios, { AxiosResponse } from 'axios';

const headers = {
  Authorization: `Bearer ${process.env.VITE_APP_TMDB_API_KEY}`,
};

const AXIOS = axios.create({
  baseURL: process.env.VITE_APP_TMDB_BASE_URL,
  timeout: 30000,
  headers,
});

/**
 * HTTP GET method, used to fetch data
 * @param {string} url - endpoint you want to reach.
 * @param {object} params - axios request params.
 */
export const getApi = (
  url: string,
  params?: Record<string, unknown>,
): Promise<AxiosResponse> =>
  new Promise((resolve, reject) => {
    AXIOS.get(url, params)
      .then(response => resolve(response))
      .catch(error => reject(error));
  });
