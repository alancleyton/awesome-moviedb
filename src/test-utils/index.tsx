import { ReactElement, ReactNode } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/config';
import { setupStore } from '@/store/config';

export type PreloadedState = Partial<RootState>;

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState;
  store?: AppStore;
}

export function renderWithProviders(
  ui: ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: { children: ReactNode }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = () => new Promise(process.nextTick);
