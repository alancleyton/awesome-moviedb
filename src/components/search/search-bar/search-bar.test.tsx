import { screen, fireEvent, waitFor, cleanup } from '@testing-library/react';
import { EnhancedStore } from '@reduxjs/toolkit';
import { AxiosResponse } from 'axios';

import { setupStore, RootState } from '@/store/config';
import { renderWithProviders, flushPromises } from '@/test-utils';
import { searchResultMock } from '@/test-utils/__mocks__';
import * as utilsApi from '@/utils/api';

import { SearchBar } from '.';

jest.mock('../../../utils/api');

describe('SearchBar', () => {
  let store: EnhancedStore<RootState>;

  beforeEach(() => {
    jest.clearAllMocks();
    store = setupStore();
  });

  afterEach(cleanup);

  it('should render correctly', () => {
    const { container } = renderWithProviders(<SearchBar />, { store });

    expect(screen.getByTestId('searchBar')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render suggestions list with search query', () => {
    renderWithProviders(<SearchBar />, { store });

    const input = screen.getByRole('searchbox');
    fireEvent.change(input, { target: { value: 'eragon' } });

    expect(screen.getByRole('listbox')).toBeInTheDocument();
  });

  it('should render loader when is searching', async () => {
    renderWithProviders(<SearchBar />, { store });

    const input = screen.getByRole('searchbox');
    fireEvent.change(input, { target: { value: 'eragon' } });

    await waitFor(
      () => {
        expect(screen.getByRole('alert')).toBeInTheDocument();
      },
      { timeout: 600 },
    );
  });

  it('should render search result when search is successfuly done', async () => {
    jest.spyOn(utilsApi, 'getApi').mockReturnValue(
      Promise.resolve({
        data: searchResultMock,
      } as AxiosResponse),
    );

    renderWithProviders(<SearchBar />, { store });

    const input = screen.getByRole('searchbox');
    fireEvent.change(input, { target: { value: 'eragon' } });

    await waitFor(
      async () => {
        await flushPromises();

        expect(
          screen.getAllByAltText(/Deadpool & Wolverine/i)[0],
        ).toBeInTheDocument();
      },
      { timeout: 600 },
    );
  });
});
