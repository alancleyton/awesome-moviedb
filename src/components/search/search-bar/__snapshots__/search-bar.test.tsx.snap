// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly 1`] = `
<div>
  <div
    class="w-full bg-white hidden"
    data-testid="searchBar"
    id="searchBar"
  >
    <div
      class="mx-auto w-full max-w-6xl p-4"
      id="searchBarForm"
    >
      <form
        action="/search"
      >
        <div
          class="a-input"
        >
          <div
            class="a-input-group a-flex a-relative a-h-fit a-w-full a-input-group-element-right"
          >
            <input
              class="a-input-field a-inline-flex a-font-sans a-font-normal a-text-gray-90 placeholder:text-gray-60 a-text-ellipsis a-leading-body a-transition a-duration-300 a-ease-in-out a-outline-none a-rounded-sm a-ring-1 enabled:hover:a-ring-2 disabled:a-text-gray-30 disabled:a-ring-gray-10 disabled:a-cursor-not-allowed a-input-field-unstyled a-ring-transparent enabled:hover:a-ring-transparent enabled:focus:a-ring-0 enabled:focus:a-ring-transparent a-input-field-md a-h-8 a-py-3 a-px-5 a-text-md a-input-field-full a-w-full a-pr-8"
              data-rac=""
              data-testid="searchField"
              id="searchField"
              name="query"
              role="searchbox"
              value=""
            />
            <div
              class="a-input-right-element a-absolute a-w-fit a-top-1/2 -a-translate-y-1/2 a-right-5"
            >
              <svg
                class="cursor-pointer"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
